zips/zip-0231.md

```diff
@@ -0,0 +1,164 @@
+The key words "MUST", "REQUIRED", "MUST NOT", "SHOULD", and "MAY" in this
+document are to be interpreted as described in BCP 14 [#BCP14]_ when, and
+only when, they appear in all capitals.
+
+
+# Abstract
+
+TBD
+
+
+# Motivation
+
+In Zcash transaction versions v2-v5, each shielded output contains a ciphertext
+comprised of a 52-byte note plaintext, and a corresponding 512-byte memo field.
+Recipients can only decrypt the outputs sent to them, and thus can also only
+observe the memo fields included with the outputs they can decrypt.
+
+The shielded transaction protocol hides the sender(s) (that is, the addresses
+corresponding to the keys used to spend the input notes) from all of the
+recipients. For certain kinds of transactions, it is desirable to make one or
+more sender addresses available to one or more recipients (for example, a reply
+address), and in subsets of those transactions it is desirable to authenticate
+the sender addresses (to give the recipient a guarantee that the address is
+controlled by a sender of the transaction). These Authenticated Reply Addresses
+require zero-knowledge proofs, and for the Orchard protocol these proofs are too
+large to fit into a 512-byte memo field.
+
+Something about light client truncation (simplifies the security argument).
+
+In the light client protocol, a recipient need not download full transaction
+information if the compact transaction information indicates that they do not
+receive any memo in the transaction.
+
+
+# Requirements
+
+- Recipients can receive memo data that is greater than 512 bytes in length.
+- Multiple recipients, across any of the shielded pools, can be given the
+  capability to view the same memo data.
+- The exact number and exact lengths of distinct decryptable memos should not
+  be revealed, even to the transaction recipients, although an upper bound on
+  the total length of memo data that the observer does not have the capability
+  to view will be leaked.
+- A recipient can determine whether or not they have been given the capability
+  to view any memo solely by decrypting the note ciphertext.
+- Memo chunks within a transaction can be individually pruned without preventing
+  the transaction from being verified.
+  - As a consequence, a recipient with the capability to view a memo can prune
+    all of that memo's chunks.
+
+
+# Non-requirements
+
+- Recipients do not need to be able to receive multiple memos per note.
+
+
+# Specification
+
+## Memo encryption
+
+Sketch:
+
+In shielded bundles, the 512-byte memo field is replaced by `[flag_byte] || memo_key`.
+
+```
+STREAM: AEAD(key, counter, pt)
+
+[
+    (key_a, 0),
+    (key_b, 0),
+    (key_a, 1),
+    (key_c, 0),
+    (key_c, 1),
+    (key_a, 2),
+]
+```
+
+## Memo bundle format
+
+```rust
+enum MemoChunk {
+    Encrypted(AEAD(MemoChunk, memo_key)),
+    Pruned {
+        memo_chunk_digest: Digest,
+    }
+}
+
+struct MemoBundle {
+    chunks: Vec<MemoChunk>,
```

Comment:
We agreed there will be a consensus limit on the number of memo chunks, which acts as a bound on the maximum contiguous memo size. We haven't yet decided on what the bound should be; @nuttycom suggested 10kiB, which would be a limit of 20 512-byte chunks (if we use that chunk size; we may want to use a smaller chunk size to have greater flexibility, at the cost of more AEAD overhead and more information revealed about candidate memo lengths).